### PR TITLE
[DPM-128] action traces

### PR DESCRIPTION
**Pull request by cyphersnake**
Store for each transaction all events and clean it before next transaction.

### CODE REVIEW

File: tester/include/game_tester/contracts.hpp.in [hunk 1]
```diff
@@ -12,18 +12,25 @@ static const eosio::chain::name events_name = N(events);
 static const eosio::chain::name casino_name = N(casino);
 
 struct contracts {
+    struct events_struct {
```
Comment (algys) [inline tester/include/game_tester/contracts.hpp.in]: it's not related to "contracts" struct, so shouldn't be under it

File: tester/include/game_tester/contracts.hpp.in [hunk 2]
```diff
@@ -12,18 +12,25 @@ static const eosio::chain::name events_name = N(events);
...
+        static fc::path folder() { return "${GAME_SDK_PATH}sdk/files/events"; }
+    };
+
     struct platform {
         struct platfrm {
+            static fc::path             abi_path() { return "${PLATFORM_CONTRACTS_PATH}/platform/platform.abi"; }
             static std::vector<uint8_t> wasm() { return read_wasm("${PLATFORM_CONTRACTS_PATH}/platform/platform.wasm"); }
             static std::vector<char>    abi() { return read_abi("${PLATFORM_CONTRACTS_PATH}/platform/platform.abi"); }
```
Comment (algys) [inline tester/include/game_tester/contracts.hpp.in]: better to use new `abi_path` in `abi` method, like: `.. abi() { return read_abi(abi_path()); }`

File: tester/include/game_tester/game_tester.hpp [hunk 3]
```diff
@@ -65,6 +65,16 @@ using EVP_PKEY_ptr = std::unique_ptr<EVP_PKEY, decltype(&::EVP_PKEY_free)>;
 
 namespace testing {
 
+enum events_id {
```
Comment (algys) [inline tester/include/game_tester/game_tester.hpp]: we prefer scoped enums

File: tester/include/game_tester/game_tester.hpp [hunk 4]
```diff
@@ -364,9 +383,95 @@ class game_tester : public TESTER {
...
+    }
+
+    void handle_transaction_ptr(const transaction_trace_ptr& transaction_trace) {
+        _events.clear();
+
+        const fc::path cpath = fc::canonical(contracts::events_struct::folder());
+        const abi_def events_abi = fc::json::from_file(contracts::platform::events::abi_path()).as<abi_def>();
+
+        abi_serializer abi_ser;
+        abi_ser.set_abi(events_abi, abi_serializer_max_time);
```
Comment (algys) [inline tester/include/game_tester/game_tester.hpp]: It's better to move things like that to another function like init_abis and call once from tester constructor, else that stuff invokes for each trx. And same with `handle_action_data`.

File: tester/include/game_tester/game_tester.hpp [hunk 5]
```diff
@@ -364,9 +383,95 @@ class game_tester : public TESTER {
...
+
+        std::for_each(
+            transaction_trace->action_traces.begin(),
+            transaction_trace->action_traces.end(),
+            [&](const auto& action_trace) {
+                if (action_trace.receiver != "events")
```
Comment (algys) [inline tester/include/game_tester/game_tester.hpp]: also add filter by action name, allow only send action

**Review by algys — CHANGES_REQUESTED** [review]

**Comment by algys** [comment]
How about this push_action: https://github.com/DaoCasino/game-contract-sdk/blob/dpm-128-action-traces/tester/include/game_tester/game_tester.hpp#L159 ? I guess, it should use our push_action to register action traces.

**Comment by algys** [comment]
Also need to add minimal usage example in tests.

**Comment by cyphersnake** [comment]
Question about [`return base_tester::push_action(std::move(act), actor);`](https://github.com/DaoCasino/game-contract-sdk/blob/dpm-128-action-traces/tester/include/game_tester/game_tester.hpp#L181). I can't find any other functions which return `transaction_ptr`. I thought all `push_` functions use out [`push_action`](https://github.com/DaoCasino/game-contract-sdk/blob/4864061ffc14a60fe8782e648eebd162c92fc4ef/tester/include/game_tester/game_tester.hpp#L192). This is't true?

**Comment by cyphersnake** [comment]
Test added

**Review by algys — APPROVED** [review]

**Pull request merged**